// redux-load-data/demo/src/app.js
import React from 'react';
import LoadData from '../../src';

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      inputUrl: 'http://ron-swanson-quotes.herokuapp.com/v2/quotes',
      url: 'http://ron-swanson-quotes.herokuapp.com/v2/quotes',
      autoRefresh: false,
    };
  }

  render() {
    const { inputUrl, url, autoRefresh } = this.state;

    return (
      <div className="container">
        <div className="section">
          <h1>LoadData Demo</h1>
          <LoadData url={url} keyPath={[url]} refreshInterval={autoRefresh ? 5000 : null} debug />
          <form
            onSubmit={(e) => {
              e.preventDefault();
              this.setState({ url: inputUrl });
            }}
          >
            <div className="columns">
              <div className="column">
                <label htmlFor="url" className="label">URL: </label>
                <input
                  className="input"
                  id="url"
                  value={inputUrl}
                  onChange={e => this.setState({ inputUrl: e.target.value })}
                />
                <button type="submit" className="button">Go</button>
                <p>Example API to test:</p>
                <ul>
                  <li>
                    <a
                      href={null}
                      onClick={() =>
                        this.setState({
                          url: 'http://ron-swanson-quotes.herokuapp.com/v2/quotes',
                          inputUrl: 'http://ron-swanson-quotes.herokuapp.com/v2/quotes',
                        })}
                    >
                      Ron Swanson Quotes
                    </a>
                  </li>
                  <li>
                    <a
                      href={null}
                      onClick={() =>
                        this.setState({
                          url: 'https://jsonplaceholder.typicode.com/posts',
                          inputUrl: 'https://jsonplaceholder.typicode.com/posts',
                        })}
                    >
                      JSON placeholder posts
                    </a>
                  </li>
                  <li>
                    <a
                      href={null}
                      onClick={() =>
                        this.setState({
                          url: 'https://jsonplaceholder.typicode.com/users',
                          inputUrl: 'https://jsonplaceholder.typicode.com/users',
                        })}
                    >
                      JSON placeholder users
                    </a>
                  </li>
                  <li>
                    <a
                      href={null}
                      onClick={() =>
                        this.setState({
                          url: 'https://jsonplaceholder.typicode.com/albums',
                          inputUrl: 'https://jsonplaceholder.typicode.com/albums',
                        })}
                    >
                      JSON placeholder albums
                    </a>
                  </li>
                </ul>
              </div>
              <div className="column">
                <label htmlFor="refreshInterval" className="label">
                  Auto Refresh (5 sec):{' '}
                </label>
                <input
                  id="refreshInterval"
                  className="checkbox"
                  checked={autoRefresh}
                  type="checkbox"
                  onChange={() => this.setState(() => ({ autoRefresh: !autoRefresh }))}
                />
              </div>
            </div>

          </form>
        </div>
      </div>
    );
  }
}

export default App;
